### PR TITLE
Implement #159 -- Semantic validation of `min_transfer_time` field from `transfers.txt`

### PR DESCRIPTION
**Pull request by lionel-nj**
**Summary:**

This PR provides support to semantically validate the value of field `min_transfer_time`.

**Expected behavior:** 

At this step of valdiation it is assumed that non null values of `min_transfer_time` are positive integers. To reach feature parity with the existing Python validator, we implement a rule checking that said field is within a certain range. 
This PR makes the range use configurable, so that any user can specify the min and max values acceptable for `min_transfer_time`.

If these threshold are not specified default values will be used:
- min range: 0 sec
- max range : 86400 sec (24 h)

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [x] Linked all relevant issues
- ~[ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)~

### CODE REVIEW

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java [hunk 1]
```diff
@@ -235,9 +235,12 @@ public String getExecParamValue(final String key) throws IllegalArgumentExceptio
             }
 
             case EXCLUSION_KEY: {
-                return hasExecParamValue(EXCLUSION_KEY) ?
-                        getExecParamByKey(EXCLUSION_KEY).getValue().toString()
-                        : null;
+                return hasExecParamValue(EXCLUSION_KEY) ? getExecParamByKey(EXCLUSION_KEY).getValue().toString() : null;
+            }
+
+            case TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN :
```
Comment (barbeau) [inline adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java:241]: Doesn't look like you've added the rule description yet, but I'd definitely include the default range values in the rule description, as well as a message saying they can be configured an a hyperlink to the documentation describing how ranges can be configured.

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java [hunk 2]
```diff
@@ -269,24 +272,10 @@ public Options getOptions() {
...
-        options.addOption(PATHWAY_MAX_SLOPE_KEY, PATHWAY_MAX_LENGTH_KEY, true,
-                "Maximum admissible value for field slope of file pathways.txt");
-        options.addOption(PATHWAY_MIN_WIDTH_LOWER_BOUND_KEY, PATHWAY_MIN_WIDTH_LOWER_BOUND_KEY, true,
-                "Minimum admissible value for field min_width of file pathways.txt");
-        options.addOption(PATHWAY_MIN_WIDTH_UPPER_BOUND_KEY, PATHWAY_MIN_WIDTH_UPPER_BOUND_KEY, true,
-                "Maximum admissible value for field min_width of file pathways.txt");
+        options.addOption(TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN, TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN, true,
+                "Minimum admissible value for field min_transfer_time of file transfers.txt");
+        options.addOption(TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX, TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX, true,
+                "Maximum admissible value for field min_transfer_time of file transfers.txt");
```
Comment (barbeau) [inline adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java:278]: I'd suggest a different word than `admissible` - maybe `allowed`?

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java [hunk 3]
```diff
@@ -422,6 +422,19 @@ public Transfer getTransferByStopPair(final String fromStopId, final String toSt
...
     }
 
+    /**
+     * Return an immutable collection of {@code Transfer} objects representing all the rows from transfers.txt
+     *
+     * @return an immutable collection of {@link Transfer} objects representing all the rows from transfers.txt
+     */
+    @Override
+    public Collection<Transfer> getTransferAll() {
+        final Set<String> fromStopIdList = transferPerStopPair.keySet();
```
Comment (barbeau) [inline adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java:432]: Why not just return an unmodifiable `Set` here? We're still instantiating an entirely new `List` data structure here which is taking up space.

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Use case execution method: checks if every record of file `transfers.txt` having a non null value for field
+     * min_transfer_time are within a certain range. This range is specified by the user of the software by execution
+     * parameters that are passed as command line arguments or defined within file `execution-parameter.json`.
+     * If this range is not specified, the use case will consider default values for said thresholds:
+     * - range min: 0 sec
+     * - range max: 86400 sec (24 hours)
+     */
+    public void execute() {
+        logger.info("Validating rule 'W009 - records from `transfers.txt` must not contain suspicious value for field "+
```
Comment (barbeau) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java:67]: I would suggest re-wording this rule - saying it's "suspicious" without context doesn't really give you a good idea of what the problem is. Maybe:
"W009 - `min_transfer_time` in `transfers.txt` is outside allowed range"

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        final int minTransferTimeRangeMin =
+                Integer.parseInt(
+                        execParamRepo.getExecParamValue(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN));
+        final int minTransferTimeRangeMax =
+                Integer.parseInt(
+                        execParamRepo.getExecParamValue(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX));
+        dataRepo.getTransferAll()
+                .forEach(transfer -> {
+                    if (transfer.getMinTransferTime() != null &&
+                            (transfer.getMinTransferTime() < minTransferTimeRangeMin ||
```
Comment (barbeau) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java:78]: In the rule documentation you should say that the max and min values are inclusive.

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java (hunk 3 above)
Comment (barbeau) [reply adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java:432]: Sorry, this should have read "unmodifiable `Map`"

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        final int minTransferTimeRangeMax =
+                Integer.parseInt(
+                        execParamRepo.getExecParamValue(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX));
+        dataRepo.getTransferAll()
+                .forEach((fromStopId, transferCollection) -> transferCollection.forEach((toStopId, transfer) ->  {
+                    if (transfer.getMinTransferTime() != null &&
+                            (transfer.getMinTransferTime() < minTransferTimeRangeMin ||
+                                    transfer.getMinTransferTime() > minTransferTimeRangeMax)) {
+                        resultRepo.addNotice(
+                                new SuspiciousMinTransferTimeNotice(minTransferTimeRangeMin, minTransferTimeRangeMax,
```
Comment (lionel-nj) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java:80]: Wondering if it wouldn't be easier to just use the `IntegerValueOutOfRangeNotice` here?

Comment (maximearmstrong) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java]: I would go with a custom notice as you did since the error is that `min_transfer_time` isn't right.

Comment (barbeau) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java]: I agree with using a custom notice. Looking at other uses of `IntegerValueOutOfRangeNotice`, they are clear errors (e.g., value less than 0), where as this is a warning based on a heuristic, where there could be valid values outside the chosen range.

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValueTest.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        verify(mockExecParamRepo, times(1)).getExecParamValue(
+                ArgumentMatchers.eq(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN));
+        verify(mockExecParamRepo, times(1)).getExecParamValue(
+                ArgumentMatchers.eq(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX));
+
+        verify(mockDataRepo, times(1)).getTransferAll();
+        // suppress warning regarding ignored result of method since it is not necessary here.
+        //noinspection ResultOfMethodCallIgnored
+        verify(mockTransfer, times(3)).getMinTransferTime();
+
```
Comment (maximearmstrong) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValueTest.java:211]: I would add a `verifyNoInteraction(MockResultRepo)` to make it explicit

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValueTest.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        verify(mockExecParamRepo, times(1)).getExecParamValue(
+                ArgumentMatchers.eq(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN));
+        verify(mockExecParamRepo, times(1)).getExecParamValue(
+                ArgumentMatchers.eq(ExecParamRepository.TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX));
+
+        verify(mockDataRepo, times(1)).getTransferAll();
+        // suppress warning regarding ignored result of method since it is not necessary here.
+        //noinspection ResultOfMethodCallIgnored
+        verify(mockTransfer, times(1)).getMinTransferTime();
+
```
Comment (maximearmstrong) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValueTest.java:253]: Same here for `verifyNoInteraction(MockResultRepo)`

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Use case execution method: checks if every record of file `transfers.txt` having a non null value for field
+     * min_transfer_time are within a certain range. This range is specified by the user of the software by execution
+     * parameters that are passed as command line arguments or defined within file `execution-parameter.json`.
+     * If this range is not specified, the use case will consider default values for said thresholds:
+     * - range min: 0 sec
+     * - range max: 86400 sec (24 hours)
+     */
+    public void execute() {
+        logger.info("Validating rule 'W009 - `min_transfer_time` is outside allowed range" + System.lineSeparator());
```
[the remark this answers is not in the document]
Comment (lionel-nj) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java:67]: Could you point me to this PR please? Wasn't that about formatting the `.json` output file: because we do not want to format this output file?

[the remark this answers is not in the document]
Comment (maximearmstrong) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java:67]: Here is @fabrice-v comment https://github.com/MobilityData/gtfs-validator/pull/294#discussion_r451166094

[the remark this answers is not in the document]
Comment (barbeau) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateMinTransferTimeValue.java:67]: I agree with @lionel-nj - newlines are ok in system logs, but not in JSON output.

File: ADDING_NEW_RULES.md [hunk 10]
```diff
@@ -84,6 +84,342 @@ public class CannotParseDateNotice extends ErrorNotice {
...
 ~~~
 
+### III - a. Configure numeric ranges
+
+Ranges for validation of numeric fields can be configured via command line arguments or within file `execution-parameter.json`.
+To do so, two classes have to be modified: [ExecParamRepository](https://github.com/MobilityData/gtfs-validator/blob/master/usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/port/ExecParamRepository.java) and [InMemoryExecParamRepository](https://github.com/MobilityData/gtfs-validator/blob/master/adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java).
+
+1. Add a key in the form of a static String in [ExecParamRepository](https://github.com/MobilityData/gtfs-validator/blob/master/usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/port/ExecParamRepository.java)
+
+Here an example:
```
Comment (fabrice-v) [inline ADDING_NEW_RULES.md:94]: Add line number or any other reference to point to the relevant modification

File: ADDING_NEW_RULES.md [hunk 11]
```diff
@@ -84,6 +84,342 @@ public class CannotParseDateNotice extends ErrorNotice {
...
+    interface ExecParamParser {
+
+        Map<String, ExecParam> parse() throws IOException;
+    }
+}
+~~~
+
+2. Add a command line option in [InMemoryExecParamRepository](https://github.com/MobilityData/gtfs-validator/blob/master/adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java)
+
+Here an example:
```
Comment (fabrice-v) [inline ADDING_NEW_RULES.md:135]: Add line number or any other reference to point to the relevant modification

File: ADDING_NEW_RULES.md [hunk 12]
```diff
@@ -84,6 +84,342 @@ public class CannotParseDateNotice extends ErrorNotice {
...
+     */
+    @Override
+    public boolean isEmpty() {
+        return execParamCollection.isEmpty();
+    }
+~~~
+
+3. Define default values for said ranges in [default-execution-parameters.json](https://github.com/MobilityData/gtfs-validator/blob/master/config/src/main/resources/default-execution-parameters.json).
+
+Here an example:
```
Comment (fabrice-v) [inline ADDING_NEW_RULES.md:331]: Add line number or any other reference to point to the relevant modification

File: ADDING_NEW_RULES.md [hunk 13]
```diff
@@ -84,6 +84,342 @@ public class CannotParseDateNotice extends ErrorNotice {
...
+                return zipInputPath;
+            }
+
+            case EXCLUSION_KEY: {
+                return hasExecParamValue(EXCLUSION_KEY) ? getExecParamByKey(EXCLUSION_KEY).getValue().toString() : null;
+            }
+
+            case TRANSFER_MIN_TRANSFER_TIME_RANGE_MIN :
+            case TRANSFER_MIN_TRANSFER_TIME_RANGE_MAX : {
+                return hasExecParamValue(key) ? getExecParamByKey(key).getValue().toString() : defaultValue.get(0);
```
[the remark this answers is not in the document]
Comment (fabrice-v) [reply ADDING_NEW_RULES.md:416]: I raised an eyebrow for the same reason but I think the get(0) doesn't refer to the actual value (it's an index in a list). Maybe this code should be clarified with a comment saying as much

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java [hunk 14]
```diff
@@ -269,24 +275,10 @@ public Options getOptions() {
         // define options related to GTFS file `pathways.txt`
         options.addOption(String.valueOf(EXCLUSION_KEY.charAt(1)), EXCLUSION_KEY, true,
                 "Exclude files from semantic GTFS validation");
-        options.addOption(PATHWAY_MIN_LENGTH_KEY, PATHWAY_MIN_LENGTH_KEY, true,
```
Comment (fabrice-v) [inline adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java:272]: Why were all those removed?

File: adapter/repository/in-memory-simple/src/test/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepositoryTest.java [hunk 15]
```diff
@@ -373,6 +373,40 @@ void addTransferAndGetTransferByStopIdPairShouldReturnSameEntity() {
         assertEquals(mockTransfer03, underTest.getTransferByStopPair("stop id 0", "stop id 5"));
     }
 
+    @Test
+    void getTransferAllShouldReturnTransferCollection() {
```
Comment (fabrice-v) [inline adapter/repository/in-memory-simple/src/test/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepositoryTest.java:377]: Shall we check for immutability as well (invoking `put` should raise an exception)

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/port/ExecParamRepository.java [hunk 16]
```diff
@@ -35,15 +35,8 @@
     String URL_KEY = "url";
     String INPUT_KEY = "input";
     String EXCLUSION_KEY = "exclude";
-    String PATHWAY_MIN_LENGTH_KEY = "pathway_min_length";
```
Comment (fabrice-v) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/port/ExecParamRepository.java:38]: why were those removed?

Comment (lionel-nj) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/port/ExecParamRepository.java:38]: They were not used actually, I'll bring them back when implementing semantic validation of field `pathways.length`.

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java (hunk 14 above)
Comment (lionel-nj) [reply adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java:272]: They were not used actually, I'll bring them back when implementing semantic validation of field pathways.length.

Comment (lionel-nj) [reply adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryExecParamRepository.java:272]: Replying to https://github.com/MobilityData/gtfs-validator/pull/297#discussion_r455870497: will do so.

File: ADDING_NEW_RULES.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (lionel-nj) [reply ADDING_NEW_RULES.md:416]: Will add comment for clarification - but yes , the idea is to retrieve the first item of a list .

File: adapter/repository/in-memory-simple/src/test/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepositoryTest.java (hunk 15 above)
Comment (lionel-nj) [reply adapter/repository/in-memory-simple/src/test/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepositoryTest.java:377]: Will report this verification for all gtfs entities

**Comment by lionel-nj** [comment]
Reply to https://github.com/MobilityData/gtfs-validator/pull/297#pullrequestreview-449954519:

As discussed, the schema will be updated to avoid confusion regarding range validation. Issue and PR to be opened. Will update this comment once they exist.

**Comment by lionel-nj** [comment]
Per discussion, semantic validation of numeric fields will not be implemented yet.
